gradle-plugin/src/gradleTest/java/com/squareup/anvil/plugin/IncrementalTest.kt

```diff
@@ -447,34 +464,73 @@ class IncrementalTest : BaseGradleTest() {
+      fun mergedModules(): Set<String> {
+        val compiledAppClasses = rootProject.path.resolve("app")
+          .resolve("build/classes")
+        val compiledLibClasses = rootProject.path.resolve("lib")
+          .resolve("build/classes")
+        val classLoader =
+          URLClassLoader(
+            listOf(compiledAppClasses, compiledLibClasses).flatMap {
+              it.walkTopDown()
+                .filter { it.isDirectory }
+                .map { it.toURI().toURL() }
+            }
+              .toList()
+              .toTypedArray(),
+            this::class.java.classLoader,
+          )
+        val appComponent = classLoader.loadClass("com.squareup.test.app.AppComponent")
+        val componentAnnotation = appComponent.getAnnotation(Component::class.java)
+        return componentAnnotation.modules.map { it.java.canonicalName }.toSet()
+      }
+
+      val bindingModule = rootProject.path.resolve("lib")
         .anvilMainGenerated
-        .anvilModule
-        .resolve("com/squareup/test/app/AppComponent.kt")
+        .resolve("com/squareup/test/lib/OtherClassBindingModule.kt")
 
-      componentModuleFile shouldExistWithTextContaining """
+      bindingModule shouldExistWithTextContaining """
         @Module
-        @ContributesTo(Any::class)
-        public abstract class AppComponentAnvilModule {
+        @ContributesTo(scope = Any::class)
+        @InternalBindingMarker(
+          originClass = OtherClass::class,
+          isMultibinding = false,
+          priority = "NORMAL",
+        )
+        public interface OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule {
           @Binds
-          public abstract fun bindTypeA(otherClass: OtherClass): TypeA
+          public fun bindTypeA(real: OtherClass): TypeA
         }
       """.trimIndent()
 
+      mergedModules() shouldContainExactly setOf(
+        "com.squareup.test.lib.OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule",
+      )
+
       otherClassPath.writeText(otherClassContent("TypeB"))
 
       shouldSucceed(":app:compileJava") {
         task(":app:compileKotlin")?.outcome shouldBe TaskOutcome.SUCCESS
       }
 
-      componentModuleFile shouldExistWithTextContaining """
+      bindingModule shouldExistWithTextContaining """
         @Module
-        @ContributesTo(Any::class)
-        public abstract class AppComponentAnvilModule {
+        @ContributesTo(scope = Any::class)
+        @InternalBindingMarker(
+          originClass = OtherClass::class,
+          isMultibinding = false,
+          priority = "NORMAL",
+        )
+        public interface OtherClassAsComSquareupTestLibTypeBToKotlinAnyBindingModule {
           @Binds
-          public abstract fun bindTypeB(otherClass: OtherClass): TypeB
+          public fun bindTypeB(real: OtherClass): TypeB
         }
       """.trimIndent()
+
+      // TODO this doesn't work, which... suggests to me that IC might be broken here?
```

Comment:
It seems like this is actually a Kotlin issue.  The `.kt` file with the invalid `TypeA` binding is not present at the start of the second compile task, but KGP/Kotlinc isn't deleting is the corresponding `.class` file.  I suspect this is because of a collision with the `[...]/OtherClassBindingModule.kt` name.

I ran this test with KGP's [build reports](https://kotlinlang.org/docs/gradle-compilation-and-caches.html#build-reports) enabled.  For the second compile task, where `TypeA` isn't bound at all, it should be deleting all the `TypeA` classes.  Instead, the report includes this:

([full log](https://github.com/square/anvil/files/14532629/second_compilation-collision.txt))
```
Process SignatureChanged(fqName = com.squareup.test.lib.OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule)
Process ParentsChanged(fqName = com.squareup.test.lib.OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule)
Deleting /lib/build/classes/kotlin/main/META-INF/lib.kotlin_module on clearing cache for /lib/build/anvil/main/generated/anvil/hint/merge/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule.kt
Deleting /lib/build/classes/kotlin/main/anvil/hint/merge/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModuleKt.class on clearing cache for /lib/build/anvil/main/generated/anvil/hint/merge/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule.kt
[...]
Deleting /lib/build/classes/kotlin/main/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeBToKotlinAnyBindingModule.class on clearing cache for /lib/build/anvil/main/generated/com/squareup/test/lib/OtherClassBindingModule.kt
Deleting /lib/build/classes/kotlin/main/META-INF/lib.kotlin_module on clearing cache for /lib/build/anvil/main/generated/com/squareup/test/lib/OtherClassBindingModule.kt
```


It's deleting the `TypeA` hint's `.class` file because of the `TypeA` hint's `.kt` file, but it's deleting the `TypeB` module `.class` file because `OtherClassBindingModule.kt` changed, and it's ignoring the `TypeA` module's `.class` file.  It then creates the `TypeB` module `.class` file again, which is how we wind up with both.

This wasn't a problem before your changes because that file's contents didn't actually change.  Coincidentally, all the changing `.class` files were linked to uniquely-named `.kt` files.

So, I forked your branch and updated the generators so that they create one file per module, and the file name matches the module: [diff](https://github.com/ZacSweers/anvil/compare/z/ksp/newBindsModule...square:anvil:rick/z/ksp/newBindsModule-individual-files).

([full log](https://github.com/square/anvil/files/14532636/second_compilation-fixed.txt))
```
Deleting /lib/build/classes/kotlin/main/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule.class on clearing cache for /lib/build/anvil/main/generated/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule.kt
Deleting /lib/build/classes/kotlin/main/META-INF/lib.kotlin_module on clearing cache for /lib/build/anvil/main/generated/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule.kt
[...]
Process SignatureChanged(fqName = com.squareup.test.lib.OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule)
Process ParentsChanged(fqName = com.squareup.test.lib.OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule)
Deleting /lib/build/classes/kotlin/main/META-INF/lib.kotlin_module on clearing cache for /lib/build/anvil/main/generated/anvil/hint/merge/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule.kt
Deleting /lib/build/classes/kotlin/main/anvil/hint/merge/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModuleKt.class on clearing cache for /lib/build/anvil/main/generated/anvil/hint/merge/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule.kt
[...]
Deleting /lib/build/classes/kotlin/main/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeBToKotlinAnyBindingModule.class on clearing cache for /lib/build/anvil/main/generated/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeBToKotlinAnyBindingModule.kt
Deleting /lib/build/classes/kotlin/main/META-INF/lib.kotlin_module on clearing cache for /lib/build/anvil/main/generated/com/squareup/test/lib/OtherClassAsComSquareupTestLibTypeBToKotlinAnyBindingModule.kt
```

So, it still deletes the brand new `TypeB` `.class` files and recreates them, but the important part is that it deletes that lingering `TypeA` `.class` file and it doesn't come back.

But your `mergedModules()` function is still finding an `OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule` reference somewhere, now with an exception when it tries to access the `Component.modules` field:

```
java.lang.ClassNotFoundException: com.squareup.test.lib.OtherClassAsComSquareupTestLibTypeAToKotlinAnyBindingModule
[...]
```

I really don't know why.  The generated stubs don't reference it, I don't see it in any decompiled .class files, and I can't find it anywhere with kotlinx-metadata.

The code compiles without issue.  I even turned the `:app` module into an application, with a `main` function that initializes the component and injects that `TypeB` dependency, and that code executes without issue.

Reply:
Very strange. Let me see if I can get a couple eyes from the compiler/KGP side to see if this is a bug we should file for the former. In the meantime, do you think we should switch to the 1:1 approach from your fork? I can try digging in on the phantom binding module ref

Reply:
Looks like things are green now with the switch to 1:1 generation 🥳

Reply:
LGTM!